index.bs

```diff
@@ -371,28 +404,175 @@ model]]:
+        [=top-level traversable/bounce tracking record=]'s
+        [=bounce tracking record/storage access set=].
+
+Issue: TODO: Handle subresource requests and iframes for client-side redirects.
+
+</div>
+
+Note: We currently don't treat cookie reads as stateful, but this would be a
+reasonable future change. We could run [=process a network cookie access for bounce tracking mitigations=]
+in the <a spec="fetch">HTTP-network-or-cache fetch</a> algorithm after step 8.21.1.2,
+"... [=append=] (`Cookie`, cookies) to httpRequest’s [=header list=]. ..."
+
+<h5 id="bounce-tracking-mitigations-response-received-monkey-patch">Response Received Monkey
+Patch</h5>
+
+When the response is received at the end of a navigation, fill the [=bounce tracking record/bounce set=].
+
+Insert the following steps in the <a spec="html">create navigation params by fetching</a> algorithm, after Step 19.7,
+"Wait until either response is non-null...", which is the point that [=response/url list=] becomes available.
+
+1. Run [=process response received for bounce tracking=] given
+    <var ignore>navigable</var> and <var ignore>response</var>'s <var ignore>URL list</var>.
+
+<div algorithm>
+
+To <dfn>process response received for bounce tracking</dfn> given a [=navigable=]
+|navigable| and a [=list=] of [=URLs=] |URLs|, perform the following steps:
+
+1. If |navigable| is not a [=top-level traversable=], then abort these steps.
+1. [=Assert=]: |navigable|'s [=top-level traversable/bounce tracking record=] is not null.
+1. Let |global| be |navigable|'s [=navigable/active document=]'s [=relevant global object=].
+1. [=Run steps after a timeout=] given:
+    <dl>
+        <dt><var ignore>global</var></dt>
+        <dd>|global|</dd>
+        <dt><var ignore>milliseconds</var></dt>
+        <dd>[=client bounce detection timer period=]</dd>
+        <dt><var ignore>completionSteps</var></dt>
+        <dd>[=queue a global task=] on the [=networking task source=] with |global| to [=record stateful bounces for bounce tracking=] given |global|</dd>
+    </dl>
+1. [=list/For each=] |URL| in |URLs|:
+    1. Let |site| be the result of running [=obtain a site=] given |URL|.
+    1. Let |host| be the |site|'s [=host=].
+    1. [=Insert=] |host| to the |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/bounce set=].
+
+</div>
+
+<h5 id="bounce-tracking-mitigations-document-load-monkey-patch">Document Loaded Monkey
+Patch</h5>
+
+When the document is loaded at the end of a navigation, update the [=bounce tracking record/final host=].
+
+Note: The [=bounce tracking record/final host=] is updated later than [=process response received for bounce tracking=]
+to ensure that the [=bounce tracking record/final host=] is a valid user destination. This is the one that has to be
+compared against when deciding to exempt a [=host=] in the [=record stateful bounces for bounce tracking=] algorithm.
```

Comment:
Interesting, so this catches the case that the final document redirects to a resource that'll set cookies but fail to load as a document?

I think we'd benefit from a state machine diagram of some sort for all this ... in a separate PR.